### PR TITLE
Implicit account transition

### LINKED ISSUE #1350: CLI: add AccountOutput bootstrapping

### LINKED ISSUE #1533: Add transition from implicit account to account

### PR DESCRIPTION
**Pull request by thibault-martinez**
Fixes #1533 & #1350

### CODE REVIEW

**Comment by Alex6323** [comment]
how can I unlook now?

File: bindings/nodejs/lib/types/wallet/bridge/account.ts
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/wallet/bridge/account.ts:1]: Is this file still needed ?

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/wallet/bridge/account.ts:1]: Should be removed and the methods should be all in wallet.ts

File: sdk/src/wallet/operations/transaction/account.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                BlockIssuerKeys::from_vec(vec![BlockIssuerKey::from(Ed25519BlockIssuerKey::from(public_key))])?,
+            )?])
+            .finish_output()?;
+
+        let transaction_options = TransactionOptions {
+            custom_inputs: Some(vec![*output_id]),
+            ..Default::default()
+        };
+
+        let prepared_transaction = self
```
Comment (Thoralf-M) [inline sdk/src/wallet/operations/transaction/account.rs]: There should be a method that returns this, so signing can be done independent
The Rust bindings should also only expose the prepare method and all in one can be done on bindings side as we do for all the other transaction methods

Comment (thibault-martinez) [reply sdk/src/wallet/operations/transaction/account.rs]: Ah, right

Comment (thibault-martinez) [reply sdk/src/wallet/operations/transaction/account.rs]: Done

File: bindings/python/iota_sdk/wallet/account.py [hunk 2]
```diff
@@ -278,6 +278,24 @@ def implicit_account_creation_address(self) -> str:
...
+    def prepare_implicit_account_transition(
+            self, output_id: OutputId) -> PreparedTransaction:
+        """Prepares to transition an implicit account to an account.
+        """
+        prepared = self._call_account_method(
+            'implicitAccountTransition', {
+                'outputId': output_id
+            }
+        )
+        return PreparedCreateTokenTransaction(
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/wallet/account.py]:
```suggestion
        return PreparedTransaction(
```

File: sdk/src/client/secret/mod.rs [hunk 3]
```diff
@@ -553,7 +553,7 @@ where
                 // We can only sign ed25519 addresses and block_indexes needs to contain the account or nft
                 // address already at this point, because the reference index needs to be lower
                 // than the current block index
-                if !input_address.is_ed25519() {
+                if !input_address.is_ed25519() && !input_address.is_implicit_account_creation() {
```
Comment (Thoralf-M) [inline sdk/src/client/secret/mod.rs]: Also want to add restricted addresses?

Comment (thibault-martinez) [reply sdk/src/client/secret/mod.rs]: Sure

File: sdk/src/wallet/operations/transaction/account.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            if implicit_account_data.output.is_implicit_account() {
+                implicit_account_data.output.as_basic()
+            } else {
+                return Err(Error::ImplicitAccountNotFound);
+            }
+        } else {
+            return Err(Error::ImplicitAccountNotFound);
+        };
+
+        let public_key = if let Some(bip_path) = self.bip_path().await {
```
Comment (Thoralf-M) [inline sdk/src/wallet/operations/transaction/account.rs:45]: I think the user should be able to provide the public key to the function as parameter

Comment (Thoralf-M) [reply sdk/src/wallet/operations/transaction/account.rs:45]: Probably somewhat related to the TODO issue in the else case

Comment (thibault-martinez) [reply sdk/src/wallet/operations/transaction/account.rs:45]: I guess I agree but there are more combinations we can do, I have opened an issue for it

File: sdk/src/client/secret/mod.rs (hunk 3 above)
Comment (thibault-martinez) [reply sdk/src/client/secret/mod.rs]: Done

File: bindings/python/iota_sdk/wallet/account.py (hunk 2 above)
Comment (thibault-martinez) [reply bindings/python/iota_sdk/wallet/account.py]: Whoops

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/client/api/block_builder/input_selection/mod.rs [hunk 5]
```diff
@@ -432,25 +441,27 @@ impl InputSelection {
...
+                                log::debug!("validate_transitions error {err:?}");
+                                return Err(Error::UnfulfillableRequirement(Requirement::Account(
+                                    *account_output.account_id(),
+                                )));
+                            }
+                        }
+                        Output::Basic(_) => {
+                            // TODO https://github.com/iotaledger/iota-sdk/issues/1664
+                        }
+                        _ => unreachable!(),
```
Comment (Alex6323) [inline sdk/src/client/api/block_builder/input_selection/mod.rs]: I think Chloé's preference here would be to use `panic` instead since this execution branch hasn't been ruled out prior. It may be unreachable protocol-wise, but not programatically/logically.

Comment (thibault-martinez) [reply sdk/src/client/api/block_builder/input_selection/mod.rs]: This is 100% programatically unreachable

Comment (thibault-martinez) [reply sdk/src/client/api/block_builder/input_selection/mod.rs]: Or maybe I don't understand what you mean by programatically

File: sdk/src/client/secret/mod.rs [hunk 6]
```diff
@@ -553,8 +553,10 @@ where
                 // We can only sign ed25519 addresses and block_indexes needs to contain the account or nft
                 // address already at this point, because the reference index needs to be lower
                 // than the current block index
-                if !input_address.is_ed25519() {
-                    Err(InputSelectionError::MissingInputWithEd25519Address)?;
+                match &input_address {
+                    Address::Ed25519(_) | Address::ImplicitAccountCreation(_) => {}
+                    Address::Restricted(restricted) if restricted.address().is_ed25519() => {}
+                    _ => Err(InputSelectionError::MissingInputWithEd25519Address)?,
```
Comment (Alex6323) [inline sdk/src/client/secret/mod.rs:559]: Not sure I like how it is currently. What about?
```rust
if { match &input_address {
    Address::Ed25519(_) | Address::ImplicitAccountCreation(_) => true,
    Address::Restricted(restricted) if restricted.address().is_ed25519() => true,
    _ => false
} } { ... } else { Err(InputSelectionError::MissingInputWithEd25519Address)? }
```
Maybe I would consider binding the match first to a `let is_valid_input_address`, but I'ld find that overall easier to read.

File: sdk/src/client/api/block_builder/input_selection/mod.rs (hunk 5 above)
Comment (Alex6323) [reply sdk/src/client/api/block_builder/input_selection/mod.rs]: like ... could it become suddenly reachable by a protocol change?

Comment (thibault-martinez) [reply sdk/src/client/api/block_builder/input_selection/mod.rs]: I guess, a massive protocol change that would break the whole codebase. I don't think that's a valid argument anyway, any change to the match could make an unreachable reachable 🤷🏻

Comment (thibault-martinez) [reply sdk/src/client/api/block_builder/input_selection/mod.rs]: I'll change it anyway as I don't really care but this is definitely unreachable

Comment (thibault-martinez) [reply sdk/src/client/api/block_builder/input_selection/mod.rs]: Changed

File: sdk/src/client/secret/mod.rs (hunk 6 above)
Comment (thibault-martinez) [reply sdk/src/client/secret/mod.rs:559]: Honestly think my version is more readable, same thing, less verbose 🤷🏻‍♂️

Comment (Alex6323) [reply sdk/src/client/secret/mod.rs:559]: okay you can keep it, but I think code should be as explicit as possible and this is essentially a somewhat hidden `if else`.

**Review by Alex6323 — APPROVED** [review]

File: sdk/src/client/secret/mod.rs (hunk 6 above)
Comment (thibault-martinez) [reply sdk/src/client/secret/mod.rs:559]: I mean, literally all match are hidden if/else, that's syntactic sugar to make things shorter/prettier
